README.md
[![Go Report Card](https://goreportcard.com/badge/github.com/Luzifer/ediplug_ctrl)](https://goreportcard.com/report/github.com/Luzifer/ediplug_ctrl)
![](https://badges.fyi/github/license/Luzifer/ediplug_ctrl)
![](https://badges.fyi/github/downloads/Luzifer/ediplug_ctrl)
![](https://badges.fyi/github/latest-release/Luzifer/ediplug_ctrl)

# Luzifer / ediplug\_ctrl

`ediplug_ctrl` is a small webserver to wrap some amount of [EdiPlug SP2101W](http://www.edimax.com/edimax/merchandise/merchandise_detail/data/edimax/au/home_automation_smart_plug/sp-1101w/) smart plugs. It is capable of fetching metrics from those plugs and to set the state through a simple API instead of messing with XML on the controlling side.

## Usage

### Starting

```bash
# ediplug_ctrl --help
Usage of ./ediplug_ctrl:
      --ip=[]: IPs of plugs to monitor / control
      --listen=":3000": Address to listen on for HTTP interface
      --password="1234": Password of the plugs
      --poll=10: Poll every N seconds
      --version[=false]: Show version and exit
```

You can run the `ediplug_ctrl` using docker or as a [single binary](https://github.com/Luzifer/ediplug_ctrl/releases):

```bash
# docker run luzifer/ediplug_ctrl --ip=10.0.0.111

# ./ediplug_ctrl --ip=10.0.0.111
```

### API

The API exposes following methods:

- `/metrics` - Metrics endpoint to be fetched by a Prometheus instance
- `/switch/<plug name>/<on/off>` - Control the plug via its name you set in the settings


History.md
# 0.3.2 / 2017-11-05

  * Add automated Github publishing
  * Update Dockerfile
  * Update README

# 0.3.1 / 2017-11-05

  * Update dependencies
  * Switch to dep for vendoring

# 0.3.0 / 2017-11-05

  * Implement Digest auth for firmware >=2.08
  * Improve logging

# 0.2.1 / 2017-03-03

  * Fix: Add locking to plug status store

# 0.2.0 / 2016-11-13

  * Add plug status API call


0.1.3 / 2016-04-22
==================

  * Fix: Maybe test the code before committing

0.1.2 / 2016-04-22
==================

  * Retry switch commands in case plug is not available

0.1.1 / 2016-04-10
==================

  * Added comments for exported resources (golint)

0.1.0 / 2016-04-10
==================

  * Initial version


ediplug/commands.go
package ediplug

import (
	"bytes"
	"encoding/xml"
	"fmt"
	"io"
	"net/http"
	"time"

	httpHelper "github.com/Luzifer/go_helpers/http"
	"golang.org/x/net/html/charset"
)

// Command is an interface wich includes a XML generator and a XML parser
type Command interface {
	GetXML() ([]byte, error)
	Parse(in io.Reader) error
}

// ExecuteCommand sends a command to a plug and parses the answer. The response can be found in the command object
func ExecuteCommand(c Command, ip string, password string) error {
	bodyRaw, err := c.GetXML()
	if err != nil {
		return err
	}

	body := bytes.NewBuffer(append([]byte(xml.Header), bodyRaw...))
	uri := fmt.Sprintf("http://%s:10000/smartplug.cgi", ip)
	req, _ := http.NewRequest("POST", uri, body)
	req.Header.Set("Content-Type", "application/xml")

	// Basic Auth is used in firmware <2.08
	req.SetBasicAuth("admin", password)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}

	if resp.StatusCode == http.StatusUnauthorized && resp.Header.Get("Www-Authenticate") != "" {
		resp.Body.Close()
		// Most likely we do have a >=2.08 firmeware here which switched to digest auth
		body := bytes.NewBuffer(append([]byte(xml.Header), bodyRaw...))
		req, _ = http.NewRequest("POST", uri, body)
		req.Header.Set("Content-Type", "application/xml")
		req.Header.Set("Authorization", httpHelper.GetDigestAuth(resp, "POST", req.URL.Path, "admin", password))
		resp, err = http.DefaultClient.Do(req)
	}

	if err != nil {
		return err
	}
	defer resp.Body.Close()

	return c.Parse(resp.Body)
}

// SetStateCommand switches a plug on or off. DesiredState needs to be set to "ON" or "OFF" (in caps)
type SetStateCommand struct {
	DesiredState string
	Success      bool
}

// GetXML assembles the request XML
func (s *SetStateCommand) GetXML() ([]byte, error) {
	c := struct {
		XMLName xml.Name `xml:"SMARTPLUG"`
		ID      string   `xml:"id,attr"`
		Command struct {
			ID    string `xml:"id,attr"`
			State string `xml:"Device.System.Power.State"`
		} `xml:"CMD"`
	}{}
	c.ID = "edimax"
	c.Command.ID = "setup"
	c.Command.State = s.DesiredState

	return xml.Marshal(c)
}

// Parse parses data from the response XML
func (s *SetStateCommand) Parse(in io.Reader) error {
	c := struct {
		XMLName xml.Name `xml:"SMARTPLUG"`
		ID      string   `xml:"id,attr"`
		Command string   `xml:"CMD"`
	}{}

	decoder := xml.NewDecoder(in)
	decoder.CharsetReader = charset.NewReaderLabel
	if err := decoder.Decode(&c); err != nil {
		return err
	}

	s.Success = c.Command == "OK"
	return nil
}

// GetStateCommand retrieves the current state from the plug (CurrentState will be "ON" or "OFF")
type GetStateCommand struct {
	CurrentState string

	comm struct {
		XMLName xml.Name `xml:"SMARTPLUG"`
		ID      string   `xml:"id,attr"`
		Command struct {
			ID    string `xml:"id,attr"`
			State string `xml:"Device.System.Power.State"`
		} `xml:"CMD"`
	}
}

// GetXML assembles the request XML
func (g *GetStateCommand) GetXML() ([]byte, error) {
	g.comm.ID = "edimax"
	g.comm.Command.ID = "get"

	return xml.Marshal(g.comm)
}

// Parse parses data from the response XML
func (g *GetStateCommand) Parse(in io.Reader) error {
	decoder := xml.NewDecoder(in)
	decoder.CharsetReader = charset.NewReaderLabel
	if err := decoder.Decode(&g.comm); err != nil {
		return err
	}

	g.CurrentState = g.comm.Command.State
	return nil
}

// GetEnergyCommand retrieves the energy stats from the plug
type GetEnergyCommand struct {
	LastToggleTime time.Time
	NowCurrent     float64 // Measured in Ampere
	NowPower       float64 // Measured in Watts
	DailyEnergy    float64 // Measured in kWh
	WeeklyEnergy   float64 // Measured in kWh
	MonthlyEnergy  float64 // Measured in kWh

	comm struct {
		XMLName xml.Name `xml:"SMARTPLUG"`
		ID      string   `xml:"id,attr"`
		Command struct {
			ID       string `xml:"id,attr"`
			NowPower struct {
				LastToggleTime string  `xml:"Device.System.Power.LastToggleTime,omitempty"`
				NowCurrent     float64 `xml:"Device.System.Power.NowCurrent,omitempty"`
				NowPower       float64 `xml:"Device.System.Power.NowPower,omitempty"`
				DailyEnergy    float64 `xml:"Device.System.Power.NowEnergy.Day,omitempty"`
				WeeklyEnergy   float64 `xml:"Device.System.Power.NowEnergy.Week,omitempty"`
				MonthlyEnergy  float64 `xml:"Device.System.Power.NowEnergy.Month,omitempty"`
			} `xml:"NOW_POWER"`
		} `xml:"CMD"`
	}
}

// GetXML assembles the request XML
func (g *GetEnergyCommand) GetXML() ([]byte, error) {
	g.comm.ID = "edimax"
	g.comm.Command.ID = "get"

	return xml.Marshal(g.comm)
}

// Parse parses data from the response XML
func (g *GetEnergyCommand) Parse(in io.Reader) error {
	decoder := xml.NewDecoder(in)
	decoder.CharsetReader = charset.NewReaderLabel
	if err := decoder.Decode(&g.comm); err != nil {
		return err
	}

	t, err := time.Parse("20060102150405", g.comm.Command.NowPower.LastToggleTime)
	if err != nil {
		return err
	}

	g.LastToggleTime = t
	g.NowCurrent = g.comm.Command.NowPower.NowCurrent
	g.NowPower = g.comm.Command.NowPower.NowPower
	g.DailyEnergy = g.comm.Command.NowPower.DailyEnergy
	g.WeeklyEnergy = g.comm.Command.NowPower.WeeklyEnergy
	g.MonthlyEnergy = g.comm.Command.NowPower.MonthlyEnergy
	return nil
}

// GetSystemInfoCommand retrieves general information about the plug
type GetSystemInfoCommand struct {
	Model           string
	FirmwareVersion string
	MacAddress      string
	SystemName      string
	DeviceTime      time.Time

	comm struct {
		XMLName xml.Name `xml:"SMARTPLUG"`
		ID      string   `xml:"id,attr"`
		Command struct {
			ID         string `xml:"id,attr"`
			SystemInfo struct {
				Model           string `xml:"Run.Model,omitempty"`
				FirmwareVersion string `xml:"Run.FW.Version,omitempty"`
				MacAddress      string `xml:"Run.LAN.Client.MAC.Address,omitempty"`
				SystemName      string `xml:"Device.System.Name,omitempty"`
			} `xml:"SYSTEM_INFO"`
			DeviceTime string `xml:"Device.System.Time"`
		} `xml:"CMD"`
	}
}

// GetXML assembles the request XML
func (g *GetSystemInfoCommand) GetXML() ([]byte, error) {
	g.comm.ID = "edimax"
	g.comm.Command.ID = "get"

	return xml.Marshal(g.comm)
}

// Parse parses data from the response XML
func (g *GetSystemInfoCommand) Parse(in io.Reader) error {
	decoder := xml.NewDecoder(in)
	decoder.CharsetReader = charset.NewReaderLabel
	if err := decoder.Decode(&g.comm); err != nil {
		return err
	}

	t, err := time.Parse("20060102150405", g.comm.Command.DeviceTime)
	if err != nil {
		return err
	}

	g.DeviceTime = t
	g.Model = g.comm.Command.SystemInfo.Model
	g.FirmwareVersion = g.comm.Command.SystemInfo.FirmwareVersion
	g.MacAddress = g.comm.Command.SystemInfo.MacAddress
	g.SystemName = g.comm.Command.SystemInfo.SystemName

	return nil
}


main.go
package main

import (
	"fmt"
	"net/http"
	"os"
	"reflect"
	"sync"
	"time"

	"github.com/Luzifer/ediplug_ctrl/ediplug"
	"github.com/Luzifer/rconfig"
	"github.com/cenkalti/backoff"
	"github.com/gorilla/mux"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/robfig/cron"
	log "github.com/sirupsen/logrus"
)

var (
	cfg = struct {
		ShowVersion  bool     `flag:"version" default:"false" description:"Show version and exit"`
		PlugIPs      []string `flag:"ip" default:"" description:"IPs of plugs to monitor / control"`
		PollInterval int      `flag:"poll" default:"10" description:"Poll every N seconds"`
		PlugPassword string   `flag:"password" default:"1234" description:"Password of the plugs"`
		Listen       string   `flag:"listen" default:":3000" description:"Address to listen on for HTTP interface"`
	}{}

	version = "dev"

	metrics        = map[string]plugMetrics{}
	plugs          = map[string]string{}
	plugStatus     = map[string]string{}
	plugStatusLock = sync.RWMutex{}

	defaultBackoff = backoff.NewExponentialBackOff()
)

type plugMetrics struct {
	Activated     prometheus.Gauge
	NowCurrent    prometheus.Gauge
	NowPower      prometheus.Gauge
	DailyEnergy   prometheus.Gauge
	WeeklyEnergy  prometheus.Gauge
	MonthlyEnergy prometheus.Gauge
}

func init() {
	rconfig.Parse(&cfg)

	if cfg.ShowVersion {
		fmt.Printf("EdiPlug Control %s", version)
		os.Exit(0)
	}

	if len(cfg.PlugIPs) == 0 || reflect.DeepEqual(cfg.PlugIPs, []string{""}) {
		rconfig.Usage()
		os.Exit(0)
	}

	defaultBackoff.MaxElapsedTime = 5 * time.Second
}

func main() {
	for _, plugIP := range cfg.PlugIPs {
		c := &ediplug.GetSystemInfoCommand{}
		if err := backoff.Retry(func() error {
			return ediplug.ExecuteCommand(c, plugIP, cfg.PlugPassword)
		}, defaultBackoff); err != nil {
			log.WithFields(log.Fields{"plug": plugIP}).WithError(err).Error("Unable to fetch system information")
			continue
		}

		commonLabels := prometheus.Labels{
			"system_name":      c.SystemName,
			"mac_address":      c.MacAddress,
			"firmware_version": c.FirmwareVersion,
			"model":            c.Model,
		}

		plugs[c.SystemName] = plugIP
		metrics[plugIP] = plugMetrics{
			Activated: prometheus.NewGauge(prometheus.GaugeOpts{
				Namespace:   "ediplug",
				Name:        "activated",
				Help:        "0 if switched off, 1 if switched on",
				ConstLabels: commonLabels,
			}),
			NowCurrent: prometheus.NewGauge(prometheus.GaugeOpts{
				Namespace:   "ediplug",
				Name:        "now_current",
				Help:        "Current in Ampere fetched last iteration",
				ConstLabels: commonLabels,
			}),
			NowPower: prometheus.NewGauge(prometheus.GaugeOpts{
				Namespace:   "ediplug",
				Name:        "now_power",
				Help:        "Power in Watt fetched last iteration",
				ConstLabels: commonLabels,
			}),
			DailyEnergy: prometheus.NewGauge(prometheus.GaugeOpts{
				Namespace:   "ediplug",
				Name:        "daily_energy",
				Help:        "Energy used within last day, measured in kWh",
				ConstLabels: commonLabels,
			}),
			WeeklyEnergy: prometheus.NewGauge(prometheus.GaugeOpts{
				Namespace:   "ediplug",
				Name:        "weekly_energy",
				Help:        "Energy used within last week, measured in kWh",
				ConstLabels: commonLabels,
			}),
			MonthlyEnergy: prometheus.NewGauge(prometheus.GaugeOpts{
				Namespace:   "ediplug",
				Name:        "monthly_energy",
				Help:        "Energy used within last month, measured in kWh",
				ConstLabels: commonLabels,
			}),
		}

		prometheus.MustRegister(metrics[plugIP].Activated)
		prometheus.MustRegister(metrics[plugIP].DailyEnergy)
		prometheus.MustRegister(metrics[plugIP].MonthlyEnergy)
		prometheus.MustRegister(metrics[plugIP].NowCurrent)
		prometheus.MustRegister(metrics[plugIP].NowPower)
		prometheus.MustRegister(metrics[plugIP].WeeklyEnergy)
	}

	fetchMetrics()

	c := cron.New()
	c.AddFunc(fmt.Sprintf("@every %ds", cfg.PollInterval), fetchMetrics)
	c.Start()

	r := mux.NewRouter()
	r.Handle("/metrics", prometheus.Handler())
	r.HandleFunc("/switch/{system}/{state}", handlePlugSwitch)
	r.HandleFunc("/status/{system}", handlePlugStatus)
	http.ListenAndServe(cfg.Listen, r)
}

func fetchMetrics() {
	for i := range cfg.PlugIPs {
		go func(plugIP string) {
			ce := &ediplug.GetEnergyCommand{}

			if err := backoff.Retry(func() error {
				return ediplug.ExecuteCommand(ce, plugIP, cfg.PlugPassword)
			}, defaultBackoff); err != nil {
				log.WithFields(log.Fields{"plug": plugIP}).WithError(err).Error("Unable to fetch metrics")
				return
			}

			metrics[plugIP].DailyEnergy.Set(ce.DailyEnergy)
			metrics[plugIP].MonthlyEnergy.Set(ce.MonthlyEnergy)
			metrics[plugIP].NowCurrent.Set(ce.NowCurrent)
			metrics[plugIP].NowPower.Set(ce.NowPower)
			metrics[plugIP].WeeklyEnergy.Set(ce.WeeklyEnergy)

			ca := &ediplug.GetStateCommand{}
			if err := backoff.Retry(func() error {
				return ediplug.ExecuteCommand(ca, plugIP, cfg.PlugPassword)
			}, defaultBackoff); err != nil {
				log.WithFields(log.Fields{"plug": plugIP}).WithError(err).Error("Unable to fetch acivation status")
				return
			}

			plugStatusLock.Lock()
			plugStatus[plugIP] = ca.CurrentState
			plugStatusLock.Unlock()

			switch ca.CurrentState {
			case "ON":
				metrics[plugIP].Activated.Set(1)
			case "OFF":
				metrics[plugIP].Activated.Set(0)
			default:
				log.WithFields(log.Fields{"plug": plugIP, "status": ca.CurrentState}).Warn("Got unexpected activation status")
			}
		}(cfg.PlugIPs[i])
	}
}

func handlePlugStatus(res http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)

	plugStatusLock.RLock()
	state, ok := plugStatus[plugs[vars["system"]]]
	plugStatusLock.RUnlock()

	if !ok {
		http.Error(res, "System not found", http.StatusNotFound)
		return
	}

	fmt.Fprintf(res, state)
}

func handlePlugSwitch(res http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)

	ip, ok := plugs[vars["system"]]
	if !ok {
		http.Error(res, "Plug not found.", http.StatusNotFound)
		return
	}

	stateRequest := &ediplug.SetStateCommand{}

	switch vars["state"] {
	case "on":
		stateRequest.DesiredState = "ON"
	case "off":
		stateRequest.DesiredState = "OFF"
	default:
		http.Error(res, "Status not possible.", http.StatusNotAcceptable)
		return
	}

	if err := backoff.Retry(func() error {
		return ediplug.ExecuteCommand(stateRequest, ip, cfg.PlugPassword)
	}, defaultBackoff); err != nil {
		http.Error(res, fmt.Sprintf("An error occurred while setting state: %s", err), http.StatusInternalServerError)
		return
	}

	http.Error(res, "OK", http.StatusOK)
}
